### ISSUE TITLE
error logging in 'writing login information to the keychain failed with error 'not enough memory resources are available to process this command'

### ISSUE THREAD

**Issue by apsoftwaredev** [microsoft/vscode#130893]
Does this issue occur when all extensions are disabled?: Yes/No
unable to start container without extensions enabled

- VS Code Version:
- Version: 1.60.0-insider (user setup)
Commit: 9dd4f31a21d63deea5e1fc322cfc802413b31d95
Date: 2021-08-16T05:21:19.431Z
Electron: 13.1.8
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.22000
- OS Version:

Steps to Reproduce:

VSCode Insiders Edition on Windows Insider gives error every time when logging in to GitHub and other sites  'writing login information to the keychain failed with error 'not enough memory resources are available to process this command'.  Also, have set up ssh, but still requires web authentication for extensions every time I open VSCode.  It gives the error, but then continues to login after the error and does not save login information.
after opening VSCode insiders
1. click login to github
2. displays error and does not save login information
3. 
3. then tested on current VSCode and doesn't require login every time and no issues with memory
(working) Version: 1.59.0 (system setup)
Commit: 379476f0e13988d90fab105c5c19e7abc8b1dea8
Date: 2021-08-04T23:13:12.822Z
Electron: 13.1.7
Chrome: 91.0.4472.124
Node.js: 14.16.0
V8: 9.1.269.36-electron.0
OS: Windows_NT x64 10.0.22000

Regular Edition does not give the error and I do not have to login to github everytime the app starts.
https://www.loom.com/share/b7b2411e8c3147a88b2fec7e1eaa03f3

**Comment by apsoftwaredev** [comment]
![image](https://user-images.githubusercontent.com/1483796/129561246-7207ec22-1514-4555-912b-34815b6ed799.png)
did not fix the issue

Snapshot of the error message:
![image](https://user-images.githubusercontent.com/1483796/129561356-e958302b-cc22-44ba-b89d-45e96661ceec.png)

**Comment by apsoftwaredev** [comment]
![image](https://user-images.githubusercontent.com/1483796/129561975-3955572a-4e1e-43e4-81c0-67ba6ead982a.png)

**Comment by TylerLeonhardt** [comment]
@apsoftwaredev there was an issue with the last insiders build. Can you update to the latest one and let us know if you're still seeing this?

**Comment by apsoftwaredev** [comment]
Thank you.
I installed the update, but unfortunately, it is still doing it.  I did a full memory diagnostic and it did not find any issues with memory, and there is plenty of it available.

Version: 1.60.0-insider (user setup)
Commit: c08913337cb8b95493be8e99d62e0b3241563f6b
Date: 2021-08-19T05:15:46.693Z
Electron: 13.1.8
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.22000

![image](https://user-images.githubusercontent.com/1483796/130110497-9fffd1c0-c1f5-4e6d-bd0c-d8e930f112e7.png)

![image](https://user-images.githubusercontent.com/1483796/130110580-fc91e906-d5d8-4b34-acbf-63dab9fd82de.png)

**Comment by apsoftwaredev** [comment]
![image](https://user-images.githubusercontent.com/1483796/130111163-7631a024-2f8c-4e88-a1cd-4a4a8358960e.png)

**Comment by apsoftwaredev** [comment]
Just updated again. Issue remains.  Thank you.

Version: 1.60.0-insider (user setup)
Commit: 89bbd372f459ce72e9c0b60eb3b26b924f64f62d
Date: 2021-08-20T17:51:50.560Z
Electron: 13.1.8
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.22000

![image](https://user-images.githubusercontent.com/1483796/130301394-c0f683ca-3f9c-4f60-8e28-51cad793169d.png)

**Comment by TylerLeonhardt** [comment]
This is likely an issue with the library we use for storing secrets called [keytar](https://github.com/atom/node-keytar) but it's impossible for me to reproduce... so it'll be hard for me to make progress here.

**Comment by marcodafonseca** [comment]
+1
I am now getting this issue on my laptop

**Comment by leonidt1721** [comment]
> +1 I am now getting this issue on my laptop
> 
> Where can I find logs or what information do you need that would assist in resolving the issue?

Hey did you manage to fix this? Same problem..

**Comment by Kazarrackt** [comment]
Third that Same issue also...

**Comment by pjlplourde** [comment]
I am also getting the same issue as well, both trying to sign in with my Microsoft Account for Settings Sync and with GitHub for Settings Sync, GitLens, or LiveShare.  Visual Studio 2022 is also having problems connecting to GitHub, and since the new connection successful screen doesn't display the token information, I cannot even connect it manually.

**Comment by TylerLeonhardt** [comment]
@apsoftwaredev do you see this only when using a Remote SSH/Remote Container connection? Your original video that it was a Remote connection.

To the others in the thread, do you _also_ experience this when using Remote?

**Comment by pjlplourde** [comment]
@TylerLeonhardt I see it when using the Accounts -> Sign In commands in the bottom left.  Whether I use it to sign in to use Settings Sync or LiveShare (with either Microsoft Account or GitHub), or using GitHub for GitLens.  I now get the token information with the instructions:

1.     Copy the token.
2.     Switch back to VS code.
3.     Click Signing in to github.com... in the status bar.
4.     Paste the token and hit enter.5. 

But, there is no "Signing in to github.com..." in the status bar, just the error message noted.

**Comment by TylerLeonhardt** [comment]
Are you using a Remote? Like Remote SSH, Remote WSL, Remote Containers?

**Comment by pjlplourde** [comment]
Thank you @PrincessRavy , that seems to have fixed it for me as well.

**Comment by karl-lunarg** [comment]
> > I solved this by changing my pagefile settings by doing: System Properties -> Advanced -> Performance -> Advanced -> Virtual Memory -> Change -> uncheck 'Automatically manage paging file size for all drives' -> check 'System managed size' -> OK -> Reboot
> > Thanks a lot it worked for me

Sadly, this did **not** work for me and I share the doubt about how a page file setting can affect this.  I even tried it with the paging file completely disabled. 

Any other ideas?

**Comment by TylerLeonhardt** [comment]
if you have nodejs installed, you can run this simple one-liner to see if the problem is how your keychain is setup:

```
npx @emacs-grammarly/keytar-cli find-creds -s vscodevscode.github-authentication
```
or if you're on Insiders:
```
npx @emacs-grammarly/keytar-cli find-creds -s vscode-insidersvscode.github-authentication
```
This is using the following open source package: https://github.com/emacs-grammarly/keytar-cli

What you should get is something like this:
```js
[
  {
    account: 'github.auth',
    password: '....... longggg string ......'
  }
]
```
> NOTE: on macOS I got a system prompt to allow this process to access the keychain... that may or may not be the case for you

If that succeeds, then the problem is possibly in vscode.

If that _does_ complain, then we'll need a fix in keytar itself.

**Comment by BladeMF** [comment]
I can try this tomorrow.

**Comment by karl-lunarg** [comment]
Hi Tyler, I understand that you can't reproduce the problem, making it hard for you to investigate.  I'm happy to try to help.

I am on Windows 11 if that makes a difference.  Also, on a completely separate Windows 10 desktop right next to me I do not have the problem. 

I ran the test you suggested and I got back the expected response for github authentication.

However, my difficulties are with microsoft authentication.  So I changed the command to look for `vscodevscode.microsoft-authentication`.

I got back 100 account/password pairs.  98 of them had live.com in the account field, for example:
```
account: 'f8cdef31-a31e-0000-93e4-5f571e91255a/1:live.com:00037FFE8EB3D2CA/f746ae30-0000-45ff-9ee7-8fe945cf5a53-66'
```
(Some digits changed in case they are sensitive)


The account fields for the other two were `microsoft.login.keylist` and `microsoft.login-1`

The password fields look quite different from the one retrurned by github.  Something like:
```
password: '{"content":"6Ekc7kbyvWHZNTyQ\\r\\n2yOxEG50lgJK6GWIsl16nNU/slY3Tl2qrtidYEQ1aQNN8zHm6ZGUXDoc3QzNqyBo\\r\\nW0
Uoxgcq54OLoqn9tG1LQzwj+iRB8vEm6hdKC+gk+KsfJkqTgTf8uP/VVeqQIVPI\\r\\nJdfm6xDtLIjXNgUpnCymOGVjjY7n4gqudT0+J
redacted
GBkQZGwMP9\\r\\nRY2P4ZoRTMzGtdPUooqoIJ3jXGhVmxaRlgBr2HZ8qe3ZihhR3W8j+42WVkPSnrfI\\r\\nNkRM3NZjUr2oyL+dD
SfnwCMkam8A0bJoOHNBdJsJzNbOEljSgAR/rY+Wi7uouDSb\\r\\nn/gWmi","hasNextChunk":true}'
```

Probably the most important thing is that only the password returned with the `microsoft.login.keylist` account looks exactly like the one returned by github, sort of like SSH keys.  The formats are the same, the digits are just different.

So, if I had to guess it looks like the caller of this function is getting back a lot more data than expected and/or needs to filter it to get only the `microsoft.login.keylist` account.

**Comment by TylerLeonhardt** [comment]
@karl-lunarg thanks for giving that a try. The _content_ of `microsoft-authentication` is different than GitHub. It has a different secret storage schema (that I could get into ... but lets call it as it is "legacy reasons")

Anyway, the fact that the script I suggested worked fine, but VS Code fails leads me to believe that this might have something to do with Electron and how `keytar` is loaded into that.

**Comment by BladeMF** [comment]
@TylerLeonhardt, I have a different problem. When I dumped my `microsoft-authentication` keychain, I have **123** entries in it so "not enough resources is probably not surprising. 

I wonder whether these accumulate over time and _then_ the problem arises. Usually, I am asked to re-authenticate once a couple of weeks so it would take a while before they accumulate.

I will remove them now so I can use my shit. 😋

**Comment by BladeMF** [comment]
For my normal VSCode (which I stopped using since we last tacked this problem), i have the following targets:
```
LegacyGeneric:target=vscodevscode.microsoft-authentication/microsoft.login.keylist
LegacyGeneric:target=vscodevscode.microsoft-authentication/microsoft.login-1
LegacyGeneric:target=vscodevscode.microsoft-authentication/f8cdef31-a31e-4b4a-93e4-5f571e91255a/1:live.com:00037FFE876F6C68/10b6fc95-334c-4b33-b4a4-33cf546ef7a7-1
LegacyGeneric:target=vscodevscode.microsoft-authentication/f8cdef31-a31e-4b4a-93e4-5f571e91255a/1:live.com:00037FFE876F6C68/eb471cf3-4937-489d-a9da-a6ced763ebe2
LegacyGeneric:target=vscodevscode.github-authentication/github.auth
LegacyGeneric:target=vscodevscode.microsoft-authentication/microsoft.login
LegacyGeneric:target=vscodevscode.microsoft-authentication/microsoft.login-3
LegacyGeneric:target=vscodevscode.microsoft-authentication/f8cdef31-a31e-4b4a-93e4-5f571e91255a/1:live.com:00037FFE876F6C68/10b6fc95-334c-4b33-b4a4-33cf546ef7a7
LegacyGeneric:target=vscodevscode.microsoft-authentication/microsoft.login-2
LegacyGeneric:target=vscodevscode.microsoft-authentication/f8cdef31-a31e-4b4a-93e4-5f571e91255a/1:live.com:00037FFE876F6C68/eb471cf3-4937-489d-a9da-a6ced763ebe2-1
```
So probably this is how it looks "in the making". Should I clean that up as well or leave it for testing?

**Comment by karl-lunarg** [comment]
FWIW, I've worked around the problem by deleting all the credentials of the form:
```
vscodevscode.microsoft-authentication/f8cdef31-a31e-4b4a-93e4-5f571e91255a/1:live.com:00037FFE8EB3D2CA/f746ae30-76e8-45ff-9ee7-8fe945cf5a53-xxx
```
where "xxx" at the end is some random-looking decimal number.

After restarting and reauthenticating, several subsequent restarts of VSCode did not require re-authentication and the number of credentials of the above form holds steady.

**Comment by pjlplourde** [comment]
I found the same thing; in my case it was a whole bunch of credentials placed there by Adobe Creative Cloud that were taking up all the space.  Once I completely cleared out the Windows Credential Store, I was able to get both VS Code and Visual Studio to correctly connect to GitHub again.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by jim-pickering** [comment]
@TylerLeonhardt That process for keychain setup worked for me, I got the result expected. So what does that mean for vscode? A bug? Every time I restart VSCode or even reload the window, I have to login again for CoPilot and Settings Sync - I use the Github login. Using MacOS Monterrey.

**Comment by drftls** [comment]
I had the same issue - too many credentials (Windows 11: Start > Credential Manager > Windows Credentials). Mine included an obscene number of vscode entries, as noted above, so I deleted them all with a powershell line I cobbled together through Google: 

`for /F "tokens=1,2 delims= " %%G in ('cmdkey /list ^| findstr vscode') do cmdkey /delete %%H`. 

It still threw the error so I hand-pruned most of the remaining entries and it took.

Good luck!

(FWIW, Win 11, 8GB RAM + 24GB Page File)
![image](https://user-images.githubusercontent.com/57203638/182243465-19af5ac0-2055-4d8c-9636-ffd1e24a9e23.png)

**Comment by szmcdull** [comment]
I think this one https://github.com/microsoft/vscode/issues/151867 is more likely the cause. Because after I deleted all the credentials in the credential manager, it still automatically logs me out. There is no out of resource error for me.

**Comment by TylerLeonhardt** [comment]
For the folks that have cleared the large list of credentials in their credential manager... can you please check to see if that number is increasing again?

My suspicion is that some of you got into a bad state around the time that I was making heavy changes in this space... but now that that has settled, vscode's code isn't polluting the cred store anymore... This is coming from my lack of reproduction of the issue even though I've tried numerous ways to do it.

With that said, users could still get into this bad state though, if they use other applications that use the keychain (like someone mentioned with Adobe products). For this case, we should at the very least include docs in the troubleshooting keychain issues here: https://code.visualstudio.com/docs/editor/settings-sync#_troubleshooting-keychain-issues

**Comment by karl-lunarg** [comment]
Hi Tyler,
I have not had any failures along the lines of the original problem (needing to login each time vscode is started) since I cleared the credentials back in July.  Today, I see only about 5 hits from the `cmdkey` command you posted.  And I think I saw about the same number when I checked at some point between July and now.  I tend to agree with your suspicion.

**Comment by szmcdull** [comment]
> ```powershell
> cmdkey /list | Select-String -Pattern "Target: LegacyGeneric:target=(vscode.+)" | ForEach-Object { cmdkey.exe /delete $_.Matches.Groups[1].Value }
>     
> ```

"Target" is localized. Need to chcp 437 first. I now clear my credentials and will try to reproduce.

**Comment by TylerLeonhardt** [comment]
Thanks @szmcdull I'll update my regex

**Comment by TylerLeonhardt** [comment]
updated mine and I'm updating your quote so that no one gets confused @szmcdull

**Comment by TylerLeonhardt** [comment]
Based on @karl-lunarg's confirmation, I will optimistically close this issue once the doc PR goes in... however, if you still see the problem of VS Code's credentials growing an unreasonable amount, please don't hesitate to reach out.

**Comment by TylerLeonhardt** [comment]
One other thing I tried is a minimal Electron repro where I created 1000 secrets to see if I could reproduce it... I could not.

For those that wanna assist in getting to the true root of the problem, I've documented a potential repro here:
https://gist.github.com/TylerLeonhardt/c1b604f0a183e3098131f29ca9176b6b

If you can reproduce the error mentioned in this issue using my Gist, please let me know so we can let the Electron team know. 

In any case, I'll still close this once https://github.com/microsoft/vscode-docs/pull/5606 goes in.

**Comment by rzhao271** [comment]
I have confirmed that once there are too many credentials, the Windows call `CredWrite` fails with `ERROR_NOT_ENOUGH_MEMORY`. This issue doesn't have to do with Electron or Node.

**Comment by TylerLeonhardt** [comment]
Given this, the best we can do is document that workaround which is done now.

**Comment by kgfly** [comment]
Hi @rzhao271 , may I ask how to list the credential? And what's the upper limit? I don't think my box have "too many credentials". But it still hit this issue. Thanks

**Comment by BladeMF** [comment]
@TylerLeonhardt , isn't it possible to remove the expired credentials automatically? It doesn't seem normal that you must delete credentials by yourself. There is not even a command that does it, you need excel/PowerShell to do it.

**Comment by kgfly** [comment]
We're still blocked by the issue. May anyone please suggest a workaround? Thanks.

**Comment by rzhao271** [comment]
@kgfly see [Accessing Credential Manager](https://support.microsoft.com/en-us/windows/accessing-credential-manager-1b5c916a-6a16-889f-8581-fc16e8165ac0) to get the list of credentials.

**Comment by kgfly** [comment]
Thanks, @rzhao271 . This is my observation: 

![image](https://user-images.githubusercontent.com/12282281/188670273-ed989f23-eeb8-49ed-93a1-ee173e9c7595.png)

I notice tons of credential created by VS code. 

To me, the bug seems: vs code **incorrectly** wrote tons of credentials to Windows, which causes the "not enough memory resources" issue. This is **not** a Windows issue but a vscode issue.

Any thoughts? Thanks

CC  @TylerLeonhardt (, who seems the PM to me)

**Comment by karl-lunarg** [comment]
I think that the working assumption is that there was a temporary problem in VS code that generated the entries in the Credential Manager.  See the explanation in this [post ](https://github.com/microsoft/vscode/issues/130893#issuecomment-1232244284)in this same issue thread.

You might try deleting those entries in your Credential Manager and make sure you have the latest revision of VS Code.  I did exactly that back in July and it worked.  I see that at least some of your entries are in or before that timeframe.

**Comment by kgfly** [comment]
Thanks. @karl-lunarg . I ran

`cmdkey /list | Select-String -Pattern "LegacyGeneric:target=(vscode.+)" | ForEach-Object { cmdkey.exe /delete $_.Matches.Groups[1].Value }`

Then the issue is gone.

I disagree with this statement:

_I have confirmed that once there are too many credentials, the Windows call CredWrite fails with ERROR_NOT_ENOUGH_MEMORY. This issue doesn't have to do with Electron or Node._

Again, to me, the bug seems: vs code incorrectly wrote tons of credentials to Windows, which causes the "not enough memory resources" issue. This is not a Windows issue but a vscode issue.

**Comment by rzhao271** [comment]
It's a Windows issue because it could be caused by other programs too. See https://github.com/microsoft/vscode/issues/130893#issuecomment-1183343570.

Adding a command to clear credentials sounds like a separate feature request to me.

**Comment by kgfly** [comment]
Hi @rzhao271 , are you saying it's by design that VS code creates tons of the credentials? 

IMO, once vscode creates ton of credentials, it's by design that Windows OS prompts OOM.

I read, https://github.com/microsoft/vscode/issues/130893#issuecomment-1183343570. in different way: both adobe/vscode should not create tons of credentials. It's their bugs to create tons of credentials.

**Comment by karl-lunarg** [comment]
I cleared my credentials back in July and haven't seen any evidence that "VS code creates tons of credentials".  Right now, I have 5 that refer to "live.com" user IDs and I suspect that has to do with my Microsoft account history.  Tyler said that was a long story and has to do with legacy/older systems.

I think I tried to explain before that Tyler suspected that there was some bad code in one of the releases of VS Code that caused the creation of these credentials.  This would have probably been in the timeframe of the first half of this year.  But apparently, that problem is no longer in current builds of VS Code.  So @kgfly, you may be right that it is a VS Code bug, but it appears to be one that is fixed, at least according to my observations.

**Comment by kgfly** [comment]
Thanks, @karl-lunarg . I think the issue still exists.

2 days ago, I have deleted all extra vscode credential using,

cmdkey /list | Select-String -Pattern "LegacyGeneric:target=(vscode.+)" | ForEach-Object { cmdkey.exe /delete $_.Matches.Groups[1].Value }

But today, login failed **again**. Then I observed that vscode creates a bunch of credential **again**, which does not sounds expected to me.

![image](https://user-images.githubusercontent.com/12282281/189053282-128b8702-e08b-4b27-b1bd-172b5facd76b.png)

@rzhao271 , do you have any idea? Thanks

**Comment by TylerLeonhardt** [comment]
> But today, login failed again. Then I observed that vscode creates a bunch of credential again, which does not sounds expected to me.

This is why I asked [up here](https://github.com/microsoft/vscode/issues/130893#issuecomment-1232244284) if anyone was seeing the large quantities of vscode secrets after clearing out their secrets.

And why I used the phrase "optimistically close" [here](https://github.com/microsoft/vscode/issues/130893#issuecomment-1233119831).

You've provided new information with that last message and you've confirmed that there's some scenario where this still happens.

With that said @kgfly can you please supply:
* if you're using a Microsoft account or work or school account (if multiple, tell me about all of them)
* how many windows of vscode you typically have open
* give me the full text of one of these keys (don't worry it's only public data)

**Comment by kgfly** [comment]
Thanks

if you're using a Microsoft account or work or school account (if multiple, tell me about all of them)
-> Only hotmail account. etc, XXX@hotmail.com

how many windows of vscode you typically have open
->
3 instance. But I use a python script to open them
`    cmd = 'code ' + path1
    if os.system(cmd) != 0:
        err_exit(cmd + 'failed.')

    cmd = 'code ' + path2
    if os.system(cmd) != 0:
        err_exit(cmd + 'failed.')

    cmd = 'code ' + path3
    if os.system(cmd) != 0:
        err_exit(cmd + 'failed.')

`

give me the full text of one of these keys (don't worry it's only public data)
->
Where to find these text?

**Comment by SmithPlatts** [comment]
> This is why I asked [up here](https://github.com/microsoft/vscode/issues/130893#issuecomment-1232244284) if anyone was seeing the large quantities of vscode secrets after clearing out their secrets.
> 
> And why I used the phrase "optimistically close" [here](https://github.com/microsoft/vscode/issues/130893#issuecomment-1233119831).
> 
> You've provided new information with that last message and you've confirmed that there's some scenario where this still happens.
> 
> With that said @kgfly can you please supply:
> 
> * if you're using a Microsoft account or work or school account (if multiple, tell me about all of them)
> * how many windows of vscode you typically have open
> * give me the full text of one of these keys (don't worry it's only public data)

Hey @TylerLeonhardt, just jumping on this as per the comment in #161220.

I have just performed the credential manager clean out as per the command line in [this comment](https://github.com/microsoft/vscode/issues/130893#issuecomment-1232244284) and my multiple VSCode windows are now all signed in, and persist across Exit's and re-open.

There were 178 entries for the specific pattern on my machine.

I will keep an eye on things, and will post back here later on.

For initial context:

* Windows Account
* 5 windows open at all times; 4 are folders/workspaces, and 1 is general files and unsaved instances
* `vscodevscode.microsoft-authentication/f8cdef31-a31e-4b4a-93e4-5f571e91255a/1:live.com:00037FFEA9CBEFA3/253ac1a7-f085-4312-a616-ede81e7a3b0c-144`

**Comment by jamgregory** [comment]
I've started experiencing the issue that appears to be mentioned in #161220 (but has now been merged into this ticket), but it appears that VSCode is correctly(?) reading my credentials from Seahorse (on Ubuntu) but still asks me to sign in every time I want to start or join a Live Share session.

I've tried clearing out all the relevant credentials from Seahorse, and that (as expected) asks me to reauthenticate with my Microsoft account, but then continues to ask me to sign in whenever I start a new session.

From Output > Microsoft Authentication, everything seems to look fine to me - it finds two(?) stored sessions and seems to refresh the token fine... :shrug: 

Details about VS Code and my OS:

> Version: 1.71.2
> Commit: 74b1f979648cc44d385a2286793c226e611f59e7
> Date: 2022-09-14T21:12:14.256Z
> Electron: 19.0.12
> Chromium: 102.0.5005.167
> Node.js: 16.14.2
> V8: 10.2.154.15-electron.0
> OS: Linux x64 5.15.0-48-generic (Ubuntu 20.04)
> Sandboxed: No

If you need to know anything else, don't hesitate to ask :+1:

**Comment by totkeks** [comment]
Just chiming in here. I just used the hint with deleting all the credentials for vscode. I had like 100+ of them. After that I signed in to settings sync using my Microsoft account. No more memory error on save. And after restarting vscode it signed me in automatically again.

**Comment by TylerLeonhardt** [comment]
I have the hardest time reproducing this issue... for those of you that do see it, can you do me a favor and:
1. Keep the Credential Manager open on the side
2. Do what you normally do (open multiple windows, just use VS Code)
3. Try to see if you can notice when the number of credentials increases and let me know if there's a particular action that causes it

I'll continue to try reproducing this...

**Comment by BladeMF** [comment]
Thank for continuing to try.
Current it seems to not happen anymore or it happens very infrequently. In credential manager I have:
* 1 github token
* 8 live.com tokens - 4 from yesterday and 4 from 30.08.2022. Not sure why 4.
* 5 microsoft.login numbered from 0 to 3 + one keylist.

**Comment by TylerLeonhardt** [comment]
@BladeMF those "8 live.com tokens" are in the form `vscodevscode.microsoft-authentication/f8cdef31-a31e-4b4a-93e4-5f571e91255a/1:live.com:00037FFEA9CBEFA3/253ac1a7-f085-4312-a616-ede81e7a3b0c-144` right?

**Comment by BladeMF** [comment]
Yes, there they are. Interestingly enough
* They were 4 pairs
* Two of the pairs are modified today (so they are somehow active, because the date changed from yesterday).

There they are (dates are in front)
```
30.08 | f8cdef31-a31e-4b4a-93e4-5f571e91255a/1:live.com:00037FFE876F6C68/10b6fc95-334c-4b33-b4a4-33cf546ef7a7
30.08 | f8cdef31-a31e-4b4a-93e4-5f571e91255a/1:live.com:00037FFE876F6C68/10b6fc95-334c-4b33-b4a4-33cf546ef7a7-1
04.11 | f8cdef31-a31e-4b4a-93e4-5f571e91255a/1:live.com:00037FFE876F6C68/3cd2a9e3-523e-4a0f-9199-59983b7a5132
04.11 | f8cdef31-a31e-4b4a-93e4-5f571e91255a/1:live.com:00037FFE876F6C68/3cd2a9e3-523e-4a0f-9199-59983b7a5132-1
30.08 | f8cdef31-a31e-4b4a-93e4-5f571e91255a/1:live.com:00037FFE876F6C68/eb471cf3-4937-489d-a9da-a6ced763ebe2
30.08 | f8cdef31-a31e-4b4a-93e4-5f571e91255a/1:live.com:00037FFE876F6C68/eb471cf3-4937-489d-a9da-a6ced763ebe2-1
04.11 | f8cdef31-a31e-4b4a-93e4-5f571e91255a/1:live.com:00037FFE876F6C68/f1a55b57-10c7-4509-9d92-1d478e7d2fc6
04.11 | f8cdef31-a31e-4b4a-93e4-5f571e91255a/1:live.com:00037FFE876F6C68/f1a55b57-10c7-4509-9d92-1d478e7d2fc6-1
```

**Comment by TylerLeonhardt** [comment]
Just a little context on the pairs... Cred Man on windows has a limit to the number of characters you can store so we chunk them. The `-1` are the 2nd part of the other one.

@BladeMF can you tell me how many things you've logged into with your Microsoft account? Based on 2 items being updated on Nov 4, I am going to guess...2... LiveShare & Settings Sync maybe?

The ones from Aug 30 are odd... around that time, I was not working heavily in credentials land so that leads me to believe that you saw the bug that other folks in this thread are mentioning... the slow slow climb in the number of secrets. It's unclear to me why that happened... I ask myself questions:

* Were these credentials being used before that date and one day they weren't read properly so new credentials were created?
  * This is possible. The library we use can be a bit dicey... and Windows certainly isn't perfect. @BladeMF before the "sign in every time behavior". Do you recall having to randomly sign in to Microsoft every once in a while? If this is the case, I need a better clean up mechanism... if possible.
* Were these credentials created accidentally as a duplicate... like could a 2nd window of VS Code think it was responsible for storing the credential and thus it stored a duplicate secret which was then overwritten by the updating of the keylist.
  * Worth double checking... I thought only one window handles the creation of secrets, but maybe I'm missing something. If this is the case, the keylist might be getting mangled with 2 instances of vscode writing to the same place

> 5 microsoft.login numbered from 0 to 3 + one keylist.

Btw, all of the non-keylist ones can probably be deleted (I'm assuming the date modified on those are a while ago). Those are from the old structure of secrets where all secrets were stored in one (chunked) credential.

**Comment by BladeMF** [comment]
> can you tell me how many things you've logged into with your Microsoft account? Based on 2 items being updated on Nov 4, I am going to guess...2... LiveShare & Settings Sync maybe?

Yes, these two.

> Do you recall having to randomly sign in to Microsoft every once in a while?

It's not that this doesn't happen, but it is diffucult to remember so far back.

**Comment by TylerLeonhardt** [comment]
Hi folks 👋 it has been a bit... I wanted to let you all know that I made a change that I think will be helpful here.

It's available in the VS Code Insiders Build: https://code.visualstudio.com/Insiders

What I'm hoping you see is no more gradually increasing number of credentials in CredMan.

Please give it a try and let me know how it goes. I know it may take days/weeks/months to fully verify that this change made a difference... but it'll be worth it!

**Comment by plwalsh** [comment]
Wish I found this issue earlier. I just counted 188 entries in my Credential Manager for `vscode-insidersvscode.microsoft-authenticator`.

I've been seeing this problem for the last few weeks (months?), where vscode constantly signs me out of my settings sync. Finally looked into it this morning and found the, "Not enough memory resources" output message, which led me to this issue.

I deleted all the authenticator tokens from CredMan, and it seems to have resolved the issue (for the last 24 hours, at least).

I use VSCode Insiders full-time and update it on a near-daily basis. Some sort of toast message, triggered by the the "not enough memory" error, might be nice to alert the user to the issue and instruct them to clear out their ms-auth tokens.

**Comment by plwalsh** [comment]
@TylerLeonhardt Spoke too soon. After a few days of daily Insiders updates and whatnot, I've continually been prompted to sign into Microsoft to sync settings, and each time, I end up with a dozen or so new tokens added into CredMan. They appear to just grow unbounded. Is that expected? Am I going to have to periodically run the `cmdkey /delete` command to clean these up? I assume I'll eventually hit the, "Not enough memory" error at some point if I just let the list continue to grow?

**Comment by TylerLeonhardt** [comment]
Thanks for the update. We'll be doing more drastic work in this space this month and next and my hope is that this will be resolved then.

**Comment by anarchodin** [comment]
I cleared out the credentials, but in less than a day I was back to more than 100 entries. The problem also started showing up on two machines that didn't have it, _immediately_ after the latest update. (As in, 100+ credentials created starting in the minutes after VS Code restarted for the upgrade.)

Notably, in light of the above comment, my setup does involve multiple restored windows - and the upgrade restart does restore them.

**Comment by TylerLeonhardt** [comment]
Thanks all. Sorry to hear this issue has not been fixed yet and sad that I've not been able to repro this. I will say that we have some plans in the works for dealing with secret storage in light of https://github.com/atom/node-keytar's archival. More on that later.

**Comment by lovettchris** [comment]
I'm also running into this, my Windows Credential manager is full of lines like this:
![image](https://user-images.githubusercontent.com/18707114/219210528-d6abcec4-dac5-4e98-a82b-1b485054fd3f.png)

See https://github.com/microsoft/vscode/issues/174290 and since yesterday it is full again, so I'd be happy to help debug this if you can point me to the right logging commands.

**Comment by szmcdull** [comment]
Not sure related or not, but the latest version would cause lsass.exe consuming high CPU, especially when VS Code thinks I need to re-log in my account.

**Comment by szmcdull** [comment]
Even after log-in, CPU usage would not restore. Must close vscode all over

**Comment by TylerLeonhardt** [comment]
I think I have finally found the issue and have a PR out for this. Thank you all for your patience and providing good information.

To help me, you can get the insider build here: https://code.visualstudio.com/insiders

My change will be in the insider build that will go out either Thursday or Friday.

**Comment by lovettchris** [comment]
So far so good, I installed the inside build, played with some different git repo folders, and I have only 2 entries in the Credential Manager:

![image](https://user-images.githubusercontent.com/18707114/221974140-f8c85a2b-ea9b-40e5-8582-b3f411c4821c.png)

The second one seems to be fore github copilot.

**Comment by TylerLeonhardt** [comment]
My change hasn't gone in yet 😅

**Comment by IllusionMH** [comment]
I haven't seen this problem for about a week on Insiders build, after several attempts to remove all credentials manually or with script. Problem re-emerged few times, but works as expected after another run of the script.

**Comment by lovettchris** [comment]
Looking at your fix, I might be more aggressive about deleting all the chunks even if one of those deletes fails, for example, could 2 vscode instances be trying to delete the same chunks at the same time?  Do you know what the user scenario is that results in very large passwords that need chunking?  That might help me test it better...

**Comment by TylerLeonhardt** [comment]
@lovettchris

> could 2 vscode instances be trying to delete the same chunks at the same time?

This isn't possible because deletes are now handled sequentially thanks to that PR.

> Do you know what the user scenario is that results in very large passwords that need chunking?

Microsoft Auth via Settings Sync. The limit in CredMan is a measly 2500 characters. A refresh token that gets stored in Cred Man is usually way larger than that.

**Comment by plwalsh** [comment]
@TylerLeonhardt Your suspected behavior written up in the PR definitely tracks with my usage patterns and observations. I'm looking forward to the fix!

Should we clear out the hundreds of tokens inside CredMan ahead of your fix going live? Or will they get cleared/handled as part of your revised operations?

Thanks!

**Comment by TylerLeonhardt** [comment]
@plwalsh you'll want to clear out CredMan.

**Comment by szmcdull** [comment]
I just updated to 1.76.1. Now multiple vscode windows cannot sign in sync, each window sign in will kick off others.

**Comment by IllusionMH** [comment]
@TylerLeonhardt I was able to reproduce on my PC with `1.76-insiders` then before update and it "spammed" credentials.

After cleaning up credentials, updating, cleaning up again - 12+ days without issues. I think I was asked once (but it might been for GHPR extension, but there is definitely no "spam" in credentials so far. (but it took noticeable time to repro it in `1.76-insiders` so hard to tell for sure)

**Comment by TylerLeonhardt** [comment]
Thanks for your report @IllusionMH. This one...... will be hard to verify.... and I think what you did is as close to a proper verification as we will get. I will mark this as verified and we will see what the 1.77 stable release gives us in a couple weeks. If anyone sees this issue in 1.77-insiders until then, please don't hesitate to reach out.

**Comment by shadowlmd** [comment]
This issue doesn't look the same as #177556, but I'll wait for 1.77 and reopen my bug report if the issue is not fixed.

**Comment by AngusMaiden** [comment]
> This issue doesn't look the same as #177556, but I'll wait for 1.77 and reopen my bug report if the issue is not fixed.

I have this exact same issue as #177556 I agree that it doesn't appear the same as #130893.

I will also wait for the new fix and see if it fixes my problem.

**Comment by BladeMF** [comment]
@TylerLeonhardt Now I am asked regularly to choose account so sync my settings from and when the list appears the Microsoft account appears as "(signed in)". I cleared the stored authentications a week ago (before the 1.77 update).

**Comment by TylerLeonhardt** [comment]
Looks like folks still see this in 1.77 at least partially. I need a new issue because this one is just way to busy... so we'll keep going in https://github.com/microsoft/vscode/issues/179202#issuecomment-1496702212

**Comment by TsingYi1263** [comment]
> 对于那些已经清除了凭证管理器中的大量凭证的人……你能检查一下这个数字是否再次增加了吗？
> 
> 我怀疑你们中的一些人在我对这个空间进行重大更改的时候陷入了糟糕的状态......但现在已经解决了，vscode 的代码不再污染 cred 存储......这即将到来尽管我已经尝试了很多方法来解决这个问题，但我没有重现这个问题。
> 
> 话虽如此，如果用户使用其他使用钥匙串的应用程序（如 Adob​​e 产品中提到的那样），用户仍然可能会进入这种不良状态。对于这种情况，我们至少应该在此处的钥匙串问题故障排除中包含文档：[https ://code.visualstudio.com/docs/editor/settings-sync#_troubleshooting-keychain-issues](https://code.visualstudio.com/docs/editor/settings-sync#_troubleshooting-keychain-issues)
> 
> ```powershell
> cmdkey /list | Select-String -Pattern "LegacyGeneric:target=(vscode.+)" | ForEach-Object { cmdkey.exe /delete $_.Matches.Groups[1].Value }
> ```

this is useful！

**Comment by yuanchen-zhu** [comment]
Just want to report and on Windows, the keychain could get into a bad state due to this issue, after which even upgrading to the newer version of vscode doesn't help'. One would need to take the actions suggested in [https ://code.visualstudio.com/docs/editor/settings-sync#_troubleshooting-keychain-issues](https://code.visualstudio.com/docs/editor/settings-sync#_troubleshooting-keychain-issues), as mentioned above, which solved the issue for me.

**Comment by BladeMF** [comment]
@TylerLeonhardt it is actually worse now. I am being asked to select the account to sync settings from repeatedly upon launching VS Code. The Microsoft accont I am using appears as "(Signed in)", so the token does not appear to be lost, just disacossiated. I can't see anything within the logs.

**Comment by paulej** [comment]
> @TylerLeonhardt it is actually worse now. I am being asked to select the account to sync settings from repeatedly upon launching VS Code. The Microsoft accont I am using appears as "(Signed in)", so the token does not appear to be lost, just disacossiated. I can't see anything within the logs.

Is it worse or did your Windows credentials just get full?  That happens to me from time-to-time and it's annoying.  [This comment](https://github.com/microsoft/vscode/issues/130893#issuecomment-1232244284) posted above provided a recipe for deleting all of the Windows credentials.  This fixes it for a while, then I have to do it again.

**Comment by TylerLeonhardt** [comment]
👋  folks, re-posting what I'm saying above:

> Looks like folks still see this in 1.77 at least partially. I need a new issue because this one is just way to busy... so we'll keep going in https://github.com/microsoft/vscode/issues/179202#issuecomment-1496702212
